FIX return proper instance class in displays classmethod

johncant opened the pull request:
At the moment, if you try this:

```
class Subclass(RocCurveDisplay):
    pass

type(Subclass.from_predictions(....))
```

You would get `RocCurveDisplay`. If users want to reuse/override functionality, like I do, and invoke by calling these named constructors, this needs to return `Subclass`.

Easily fixed by instantiating `cls` in the named constructor methods instead of i.e. RocCurveDisplay.

johncant committed:
Fix sklearn.metics.*Display.from_* use in subclass

glemaitre commented:
Yep it makes a lot of sense. I think that we can considered it as a bug.
Could you modify the other displays that suffer from the same pattern.
From what I see, the following will need to be modified as well.

- `CalibrationDisplay`
- `PartialDependenceDisplay`
- `DecisionBoundaryDisplay`

I think that we need an entry in the changelog as a FIX.
We also need a non-regression tests for these displays. We could for instance tests to subclassing is working as expected.

johncant commented:
Hi - I still plan to do this

glemaitre commented:
Cool. Thanks @johncant I am assigning the issue to you.
Do not hesitate to ping my GitHub handle when you submit the pull request.

glemaitre commented:
Ups, this is actually already a pull-request :). I will be notified when you are going to make changes, so all good.

johncant committed:
Merge branch 'main' into plotting_named_constructor_inheritance

johncant commented:
Classes to fix (from `it grep 'class [A-Za-z0-9]\+Display'`):

|Class|Named constructors need fixing|Fixed|Non-regression test
|---|-|-|-|
|CalibrationDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|DetCurveDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|PrecisionRecallDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|RocCurveDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|DecisionBoundaryDisplay:|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|PartialDependenceDisplay:|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|ConfusionMatrixDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|LearningCurveDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|ValidationCurveDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|
|PredictionErrorDisplay|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|<ul><li>- [ ] </li></ul>|

I'll update this commend as I go

johncant committed:
Fix remaining Display class named ctors + test

johncant committed:
Update changelog for #27675

johncant commented:
@glemaitre - this is what I've done:
  - Fixed all `*Display` class named constructors to return the correct type
  - Added non-regression tests for all the `*Display` classes, even those that didn't need fixing
  - Updated the changelog
  
PR is now ready!

glemaitre commented:
Thanks. I'm going to review it.

glemaitre commented on sklearn/model_selection/tests/test_plot.py:
```diff
@@ -570,3 +570,38 @@ def test_validation_curve_xscale_from_param_range_provided_as_a_list(
     assert display.ax_.get_xscale() == xscale
+
+
+def test_learning_curve_display_named_constructor_return_type(pyplot, data):
+    """Check that named constructors return the correct type when subclassed
+
+    Non-regression test for #27675
+    """
+    X, y = data
+    estimator = DecisionTreeClassifier(random_state=0)
+
+    class SubclassOfDisplay(LearningCurveDisplay):
+        pass
+
+    display = SubclassOfDisplay.from_estimator(estimator, X, y)
+
+    assert isinstance(display, SubclassOfDisplay)
+
+
+def test_validation_curve_display_named_constructor_return_type(pyplot, data):
+    """Check that named constructors return the correct type when subclassed
+
+    Non-regression test for #27675
+    """
+    X, y = data
+    estimator = DecisionTreeClassifier(random_state=0)
+    param_name, param_range = "max_depth", [1, 3, 5]
+
+    class SubclassOfDisplay(ValidationCurveDisplay):
+        pass
+
+    display = SubclassOfDisplay.from_estimator(
+        estimator, X, y, param_name=param_name, param_range=param_range
+    )
+
+    assert isinstance(display, SubclassOfDisplay)
```
We can parametrize this test:

```suggestion
@pytest.mark.parametrize(
    "Display, params", [
        (LearningCurveDisplay, {}),
        (ValidationCurveDisplay, {"param_name": "max_depth", "param_range": [1, 3, 5]}),
    ]
)
def test_subclassing_displays(pyplot, data, Display, params):
    """Check that named constructors return the correct type when subclassed.

    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/pull/27675
    """
    X, y = data
    estimator = DecisionTreeClassifier(random_state=0)

    class SubclassOfDisplay(Display):
        pass

    display = SubclassOfDisplay.from_estimator(estimator, X, y, **params)
    assert isinstance(display, SubclassOfDisplay)
```

glemaitre commented on sklearn/metrics/_plot/tests/test_common_curve_display.py:
```diff
@@ -223,3 +225,44 @@ def test_display_curve_error_pos_label(pyplot, data_binary, Display):
+def test_classifier_display_curve_named_constructor_return_type(
+    pyplot, data_binary, Display, constructor
+):
+    """Check that named constructors return the correct type when subclassed
```
```suggestion
    """Check that named constructors return the correct type when subclassed.
```

glemaitre commented on sklearn/metrics/_plot/tests/test_common_curve_display.py:
```diff
@@ -223,3 +225,44 @@ def test_display_curve_error_pos_label(pyplot, data_binary, Display):
+):
+    """Check that named constructors return the correct type when subclassed
+
+    Non-regression test for #27675
```
```suggestion
    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/pull/27675
```

glemaitre commented on sklearn/metrics/_plot/tests/test_common_curve_display.py:
```diff
@@ -223,3 +225,44 @@ def test_display_curve_error_pos_label(pyplot, data_binary, Display):
+
+    if constructor == "from_predictions":
+        curve = SubclassOfDisplay.from_predictions(y, y_pred)
+    elif constructor == "from_estimator":
```
```suggestion
    else:  # constructor == "from_estimator"
```

glemaitre commented on sklearn/inspection/_plot/tests/test_plot_partial_dependence.py:
```diff
@@ -1117,3 +1117,23 @@ def test_partial_dependence_display_with_constant_sample_weight(
+):
+    """Check that named constructors return the correct type when subclassed.
+
+    Non-regression test for #27675
```
```suggestion
    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/pull/27675
```

glemaitre commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -591,3 +591,18 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+def test_subclass_named_constructors_return_type_is_subclass(pyplot):
+    """Check that named constructors return the correct type when subclassed.
+
+    Non-regression test for #27675
```
```suggestion
    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/pull/27675
```

glemaitre reviewed (commented):
LGTM. If we make it before the release, I think we can include it in 1.4.0. Otherwise we will have to move it for 1.4.1.

johncant committed:
Update sklearn/model_selection/tests/test_plot.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

johncant committed:
Update sklearn/metrics/_plot/tests/test_common_curve_display.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

johncant committed:
Update sklearn/metrics/_plot/tests/test_common_curve_display.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

johncant committed:
Update sklearn/metrics/_plot/tests/test_common_curve_display.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

johncant committed:
Update sklearn/inspection/_plot/tests/test_boundary_decision_display.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

johncant committed:
Update sklearn/inspection/_plot/tests/test_plot_partial_dependence.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented:
I might have introduce a linter issue. You can run black as specified by the bot to solve it.

johncant committed:
Misc code formatting

johncant commented:
Hi @glemaitre - I've accepted all your suggestions and fixed the linter. LGTM!

jeremiedbb committed:
reorder what's new

jeremiedbb reviewed (approved):
LGTM. I'm +1 to include it in 1.4.0. @glemaitre you want to take another look ?

glemaitre reviewed (approved):
Fine on my side as well.
